src/mscorlib/shared/System/Globalization/TextInfo.Windows.cs

```diff
@@ -24,115 +23,29 @@ private unsafe void FinishInitialization()
             _sortHandle = ret > 0 ? handle : IntPtr.Zero;
         }
 
-        private unsafe string ChangeCase(string s, bool toUpper)
+        private unsafe void ChangeCase(char* pSource, int pSourceLen, char* pResult, int pResultLen, bool toUpper)
         {
             Debug.Assert(!_invariantMode);
-
-            Debug.Assert(s != null);
-
-            //
-            //  Get the length of the string.
-            //
-            int nLengthInput = s.Length;
-
-            //
-            //  Check if we have the empty string.
-            //
-            if (nLengthInput == 0)
-            {
-                return s;
-            }
-
-            int ret;
+            Debug.Assert(pSourceLen == pResultLen);
```

Comment:
Should we add debug asserts that check pSource, pResult is not null?

Reply:
`Debug.Assert(pSourceLen == pResultLen);` => `Debug.Assert(pSourceLen <= pResultLen);` ?

Reply:
> Should we add debug asserts that check pSource, pResult is not null?

Sure

> Debug.Assert(pSourceLen == pResultLen); => Debug.Assert(pSourceLen <= pResultLen); ?

Why?  Is there a situation where it'll be < rather than ==?

Reply:
This method is called for both strings and spans, correct?

The spans passed in to this method can come from the user who passed in a larger destination:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/MemoryExtensions.Fast.cs#L207

At the call site, the `source.Length - length` could be <= `destination.Length = length`
` ChangeCase(a, source.Length - length, b, destination.Length - length, toUpper);`

Reply:
Ah, you're right.  Will fix.